<a href="https://colab.research.google.com/github/Sharaffin/Credit-card-default-prediction/blob/main/credit_card_default_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Predicting whether a customer will default on his/her credit card </u></b>

## <b> Problem Description </b>

### This project is aimed at predicting the case of customers default payments in Taiwan. From the perspective of risk management, the result of predictive accuracy of the estimated probability of default will be more valuable than the binary result of classification - credible or not credible clients. We can use the [K-S chart](https://www.listendata.com/2019/07/KS-Statistics-Python.html) to evaluate which customers will default on their credit card payments


## <b> Data Description </b>

### <b>Attribute Information: </b>

### This research employed a binary variable, default payment (Yes = 1, No = 0), as the response variable. This study reviewed the literature and used the following 23 variables as explanatory variables:
* ### X1: Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit.
* ### X2: Gender (1 = male; 2 = female).
* ### X3: Education (1 = graduate school; 2 = university; 3 = high school; 4 = others).
* ### X4: Marital status (1 = married; 2 = single; 3 = others).
* ### X5: Age (year).
* ### X6 - X11: History of past payment. We tracked the past monthly payment records (from April to September, 2005) as follows: X6 = the repayment status in September, 2005; X7 = the repayment status in August, 2005; . . .;X11 = the repayment status in April, 2005. The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above.
* ### X12-X17: Amount of bill statement (NT dollar). X12 = amount of bill statement in September, 2005; X13 = amount of bill statement in August, 2005; . . .; X17 = amount of bill statement in April, 2005.
* ### X18-X23: Amount of previous payment (NT dollar). X18 = amount paid in September, 2005; X19 = amount paid in August, 2005; . . .;X23 = amount paid in April, 2005.

### 1. **Connection with the Data**
Here, we are provided with a dataset with 30000 customers. 
At first, we import the libraries or functions for making our analysis easy and then get connected to the set of outcomes.

In [ ]:
# Taking help of the Following Libraries
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

In [ ]:
#Mounting the drive
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [ ]:
#Creating a path in drive
path = '/content/drive/MyDrive/'

In [ ]:
#reading the dataset 
df = pd.read_csv(path + 'default of credit card clients.csv', encoding='latin-1')

# **2. Deeper Understanding of the Data**
As there is a huge no. of experiences, we take the help of statistics to measure each and every features in different dimensions and thus step by step, we find the most important features or the exact way to decode the experiences.

**“what gets measured gets done“.**

In [ ]:
#first five observations
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1    20000.0    2          2         1   24      2      2     -1     -1   
1   2   120000.0    2          2         2   26     -1      2      0      0   
2   3    90000.0    2          2         2   34      0      0      0      0   
3   4    50000.0    2          2         1   37      0      0      0      0   
4   5    50000.0    1          2         1   57     -1      0     -1      0   

   PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  \
0     -2     -2     3913.0     3102.0      689.0        0.0        0.0   
1      0      2     2682.0     1725.0     2682.0     3272.0     3455.0   
2      0      0    29239.0    14027.0    13559.0    14331.0    14948.0   
3      0      0    46990.0    48233.0    49291.0    28314.0    28959.0   
4      0      0     8617.0     5670.0    35835.0    20940.0    19146.0   

   BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0        0.0       0.0     689.0       0.0       0.0       0.0       0.0   
1     3261.0       0.0    1000.0    1000.0    1000.0       0.0    2000.0   
2    15549.0    1518.0    1500.0    1000.0    1000.0    1000.0    5000.0   
3    29547.0    2000.0    2019.0    1200.0    1100.0    1069.0    1000.0   
4    19131.0    2000.0   36681.0   10000.0    9000.0     689.0     679.0   

   default.payment.next.month  
0                           1  
1                           1  
2                           0  
3                           0  
4                           0

In [ ]:
#last five observations
df.tail()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
29995  29996   220000.0    1          3         1   39      0      0      0   
29996  29997   150000.0    1          3         2   43     -1     -1     -1   
29997  29998    30000.0    1          2         2   37      4      3      2   
29998  29999    80000.0    1          3         1   41      1     -1      0   
29999  30000    50000.0    1          2         1   46      0      0      0   

       PAY_4  PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  BILL_AMT3  BILL_AMT4  \
29995      0      0      0   188948.0   192815.0   208365.0    88004.0   
29996     -1      0      0     1683.0     1828.0     3502.0     8979.0   
29997     -1      0      0     3565.0     3356.0     2758.0    20878.0   
29998      0      0     -1    -1645.0    78379.0    76304.0    52774.0   
29999      0      0      0    47929.0    48905.0    49764.0    36535.0   

       BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  \
29995    31237.0    15980.0    8500.0   20000.0    5003.0    3047.0    5000.0   
29996     5190.0        0.0    1837.0    3526.0    8998.0     129.0       0.0   
29997    20582.0    19357.0       0.0       0.0   22000.0    4200.0    2000.0   
29998    11855.0    48944.0   85900.0    3409.0    1178.0    1926.0   52964.0   
29999    32428.0    15313.0    2078.0    1800.0    1430.0    1000.0    1000.0   

       PAY_AMT6  default.payment.next.month  
29995    1000.0                           0  
29996       0.0                           0  
29997    3100.0                           1  
29998    1804.0                           1  
29999    1000.0                           1

In [ ]:
#To get the Rows and Columns in the dataset
df.shape

(30000, 25)

In [ ]:
#information about the database
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          30000 non-null  int64  
 1   LIMIT_BAL                   30000 non-null  float64
 2   SEX                         30000 non-null  int64  
 3   EDUCATION                   30000 non-null  int64  
 4   MARRIAGE                    30000 non-null  int64  
 5   AGE                         30000 non-null  int64  
 6   PAY_0                       30000 non-null  int64  
 7   PAY_2                       30000 non-null  int64  
 8   PAY_3                       30000 non-null  int64  
 9   PAY_4                       30000 non-null  int64  
 10  PAY_5                       30000 non-null  int64  
 11  PAY_6                       30000 non-null  int64  
 12  BILL_AMT1                   30000 non-null  float64
 13  BILL_AMT2                   300

In [ ]:
#description about the dataset
df.describe()

ID   LIMIT_BAL       SEX  EDUCATION  MARRIAGE       AGE  \
count  30000.00    30000.00  30000.00   30000.00  30000.00  30000.00   
mean   15000.50   167484.32      1.60       1.78      1.55     35.49   
std     8660.40   129747.66      0.49       0.73      0.52      9.22   
min        1.00    10000.00      1.00       0.00      0.00     21.00   
25%     7500.75    50000.00      1.00       1.00      1.00     28.00   
50%    15000.50   140000.00      2.00       2.00      2.00     34.00   
75%    22500.25   240000.00      2.00       2.00      2.00     41.00   
max    30000.00  1000000.00      2.00       3.00      3.00     79.00   

          PAY_0     PAY_2     PAY_3     PAY_4     PAY_5     PAY_6  BILL_AMT1  \
count  30000.00  30000.00  30000.00  30000.00  30000.00  30000.00   30000.00   
mean      -0.02     -0.13     -0.17     -0.22     -0.27     -0.29   51223.33   
std        1.12      1.20      1.20      1.17      1.13      1.15   73635.86   
min       -2.00     -2.00     -2.00     -2.00     -2.00     -2.00 -165580.00   
25%       -1.00     -1.00     -1.00     -1.00     -1.00     -1.00    3558.75   
50%        0.00      0.00      0.00      0.00      0.00      0.00   22381.50   
75%        0.00      0.00      0.00      0.00      0.00      0.00   67091.00   
max        8.00      8.00      8.00      8.00      8.00      8.00  964511.00   

       BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  BILL_AMT6   PAY_AMT1  \
count   30000.00   3.00e+04   30000.00   30000.00   30000.00   30000.00   
mean    49179.08   4.70e+04   43262.95   40311.40   38871.76    5663.58   
std     71173.77   6.93e+04   64332.86   60797.16   59554.11   16563.28   
min    -69777.00  -1.57e+05 -170000.00  -81334.00 -339603.00       0.00   
25%      2984.75   2.67e+03    2326.75    1763.00    1256.00    1000.00   
50%     21200.00   2.01e+04   19052.00   18104.50   17071.00    2100.00   
75%     64006.25   6.02e+04   54506.00   50190.50   49198.25    5006.00   
max    983931.00   1.66e+06  891586.00  927171.00  961664.00  873552.00   

       PAY_AMT2   PAY_AMT3   PAY_AMT4   PAY_AMT5   PAY_AMT6  \
count  3.00e+04   30000.00   30000.00   30000.00   30000.00   
mean   5.92e+03    5225.68    4826.08    4799.39    5215.50   
std    2.30e+04   17606.96   15666.16   15278.31   17777.47   
min    0.00e+00       0.00       0.00       0.00       0.00   
25%    8.33e+02     390.00     296.00     252.50     117.75   
50%    2.01e+03    1800.00    1500.00    1500.00    1500.00   
75%    5.00e+03    4505.00    4013.25    4031.50    4000.00   
max    1.68e+06  896040.00  621000.00  426529.00  528666.00   

       default.payment.next.month  
count                    30000.00  
mean                         0.22  
std                          0.42  
min                          0.00  
25%                          0.00  
50%                          0.00  
75%                          0.00  
max                          1.00

There are 30,000 distinct credit card clients.

The average value for the amount of credit card limit is 167,484 with max value being 10 lakhs.

The average age is 35

Education Level is mostly graduate school and university.

Most of the clients are either married or single

In [ ]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default.payment.next.month'],
      dtype='object')

In [ ]:
#Checking missing values in each column
df.isna().sum()

ID                            0
LIMIT_BAL                     0
SEX                           0
EDUCATION                     0
MARRIAGE                      0
AGE                           0
PAY_0                         0
PAY_2                         0
PAY_3                         0
PAY_4                         0
PAY_5                         0
PAY_6                         0
BILL_AMT1                     0
BILL_AMT2                     0
BILL_AMT3                     0
BILL_AMT4                     0
BILL_AMT5                     0
BILL_AMT6                     0
PAY_AMT1                      0
PAY_AMT2                      0
PAY_AMT3                      0
PAY_AMT4                      0
PAY_AMT5                      0
PAY_AMT6                      0
default.payment.next.month    0
dtype: int64

In [ ]:
# Understand the data by checking the first row
df.loc[0]

ID                                1.0
LIMIT_BAL                     20000.0
SEX                               2.0
EDUCATION                         2.0
MARRIAGE                          1.0
AGE                              24.0
PAY_0                             2.0
PAY_2                             2.0
PAY_3                            -1.0
PAY_4                            -1.0
PAY_5                            -2.0
PAY_6                            -2.0
BILL_AMT1                      3913.0
BILL_AMT2                      3102.0
BILL_AMT3                       689.0
BILL_AMT4                         0.0
BILL_AMT5                         0.0
BILL_AMT6                         0.0
PAY_AMT1                          0.0
PAY_AMT2                        689.0
PAY_AMT3                          0.0
PAY_AMT4                          0.0
PAY_AMT5                          0.0
PAY_AMT6                          0.0
default.payment.next.month        1.0
Name: 0, dtype: float64

In [120]:
# Get the proportion of customers who had default payment in the next month (Oct.2005) 
# About 22% customers had default payment next month

df['default.payment.next.month'].value_counts(normalize=True)

0    0.78
1    0.22
Name: default.payment.next.month, dtype: float64

In [ ]:
#Get SEX column's distribution. 1: male; 2: female
df['SEX'].value_counts(dropna=False)

2    18112
1    11888
Name: SEX, dtype: int64

In [ ]:
# Get the total counts of each EDUCATION level
df.EDUCATION.value_counts()

2    14030
1    10585
3     4917
5      280
4      123
6       51
0       14
Name: EDUCATION, dtype: int64

In [ ]:
df.MARRIAGE.value_counts()

2    15964
1    13659
3      323
0       54
Name: MARRIAGE, dtype: int64

In [ ]:
df.EDUCATION.value_counts()

2    14030
1    10585
3     4917
5      280
4      123
6       51
0       14
Name: EDUCATION, dtype: int64

In [ ]:
fil =(df.EDUCATION == 5)|(df.EDUCATION == 6)|(df.EDUCATION == 0)
df.loc[fil,'EDUCATION'] = 4


In [ ]:
df['EDUCATION'].replace([0, 5,6], 4, inplace=True)

In [ ]:
df.EDUCATION.value_counts()

2    14030
1    10585
3     4917
4      468
Name: EDUCATION, dtype: int64

In [ ]:
df['EDUCATION']=np.where(df['EDUCATION'] == 4, 0, df['EDUCATION'])
df['EDUCATION']=np.where(df['EDUCATION'] == 5, 0, df['EDUCATION'])
df['EDUCATION']=np.where(df['EDUCATION'] == 6, 0, df['EDUCATION'])

In [ ]:
df["EDUCATION"].value_counts()

2    14030
1    10585
3     4917
5      280
4      123
6       51
0       14
Name: EDUCATION, dtype: int64

In [ ]:
df.PAY_0.value_counts()

 0    14737
-1     5686
 1     3688
-2     2759
 2     2667
 3      322
 4       76
 5       26
 8       19
 6       11
 7        9
Name: PAY_0, dtype: int64